### ISSUE TITLE
FR: allow more customization of uv init via env vars or settings

### ISSUE THREAD

**Issue by richieadler** [astral-sh/uv#7931]
The current defaults for `uv init` may be inconvenient, so having a way to set them via env var or {pyproject|uv}.toml would be useful.

* Assuming `--vcs git` is irksome for those of us that don't use Git in all cases.
* `--no-readme`, `--no-pin-python` and `--no-workspace` is needed even when one is creating fast test projects where those are not needed.

**Comment by zanieb** [comment]
It seems fine to allow configuring these in the `uv.toml`, I think.

**Comment by chrisrodrigue** [comment]
Would configuration in`pyproject.toml` also be possible?

**Comment by chrisrodrigue** [comment]
Related: https://github.com/astral-sh/uv/issues/5737

**Comment by zanieb** [comment]
Like.. targeting workspace members in that context? It seems like a weird use-case.

**Comment by chrisrodrigue** [comment]
More so to enable persistent uv command options and parameters for `uv init’, `uv venv`, `uv sync`, and so forth.

[Hatch](https://hatch.pypa.io/latest/intro/#configuration) and [pdm](https://pdm-project.org/latest/usage/config/#passing-constant-arguments-to-every-pdm-invocation) both support this.

The main benefit for this is so that that uv command settings and customizations are centralized in the project so that all devs working in the codebase can see and use them.

Keeping customizations in environment variables or `uv.toml` adds hidden state outside of the project scope and is not the best practice (with the exception of settings you absolutely don’t want in the repository such as credentials or secrets).

**Comment by chrisrodrigue** [comment]
I think for flexibility and consistency, an dev could/should have the ability to set every uv “configurable” via: command line, `pyproject.toml`, `uv.toml`, and environment variable, with resolution performed in that order of priority.

I don’t know what implementation would look like, but I think it could be the same abstract implementation for each `Configurable`, using the command line string as the universal key (same for command line, `pyproject.toml`, `uv.toml`, and env var) which maps to the data/function that actually modifies the command/behavior.

**Comment by sliedes** [comment]
> I think for flexibility and consistency, a dev could/should have the ability to set every uv “configurable” via: command line, `pyproject.toml`, `uv.toml`, and environment variable, with override resolution performed in that order of priority.
> 
> I don’t know what implementation would look like, but I think it could be the same abstract implementation for each `Configurable`, using the command line string as the universal key (same for command line, `pyproject.toml`, `uv.toml`, and env var) which maps to the data/function that actually modifies the command/behavior.

I think there may be two different use cases commingled here (though I may also be confused):

1. Global options that strictly only customize what `uv init` does, but *never* have an effect otherwise.

This is useful when I am opinionated about settings for projects I initialize, but I want those projects to work exactly the same on someone else's machine.

For example, suppose I generally want my projects to have some mypy option set to enable or disable some error code. I believe it would be less than ideal to have a global setting to enable or disable that; it would make mypy behave differently on the same project on this machine and other machines. Or, I might want my new projects to have a dev-dependency on mypy and ruff, but I absolutely want that dependency to be expressed in pyproject.toml.

2. Global options that are honored in projects generally (possibly unless overridden in pyproject.toml). These would likely be useful for customizations that are necessarily more for the machine/environment than the project. You could have things like local mirrors of pypi, or http proxies to use.

**Comment by chrisrodrigue** [comment]
Yeah, you got the idea. It’s a more general solution rather than solving each one off “can you please let me modify `uv xyz` via `< command line | .toml | ENV>`.”

A general solution might not be trivial. It could require careful security/performance considerations to make everything in the uv ABI configurable by every mechanism. But I bet that there there is more than one issue in this repository that could be closed. I think it would be worth it, even it was just for one thing: consistency.

I’ve found myself scratching my head, wondering why I can’t do something as simple as passing an override at the command line. For example, if I have `offline = true` set in pyproject.toml, I should be able to pass `--online` to override the setting. Alas, I cannot.

Maybe I am being a pedant, but I think that commenting a line out of my pyproject.toml or deleting an environment variable just to temporarily change a command’s behavior is an antipattern for users. Addressing every GitHub issue that requests a special case for config is an antipattern for the developers. This issue is requesting it for uv init, so case in point.
